algorithms/Path/shortest_path/tg_shortest_ss_pos_wt_tb.gsql

```diff
@@ -0,0 +1,156 @@
+ e_type: edge types to traverse                   display_edges: output edges for visualization
+ wt_attr: attribute for edge weights              result_attr: INT attr to store results to
+ wt_type: weight data type (INT,FLOAT,DOUBLE)     spit_to_file: spit data to file 
+ epsilon: min delta weight                        file_path: file to write CSV output to
+ write_size: number of paths to write concurrently
+*/
+
+  TYPEDEF TUPLE<FLOAT cost, INT length, VERTEX pred> min_tup;
+
+  HeapAccum<min_tup>(1, cost ASC, length ASC) @minPath;
+  MinAccum<FLOAT> @prevMinPath = -1;
+  ListAccum<VERTEX> @path;
+  OrAccum @is_candidate = FALSE;
+  
+  SetAccum<VERTEX> @@next_tmp;
+  ListAccum<VERTEX> @path_receiver, @path_sender;
+  FILE f(file_path);
+  UINT print_count = 0;
+   
+  # Check wt_type parameter
+	IF wt_type NOT IN ("UINT", "INT", "FLOAT", "DOUBLE") THEN
+		PRINT "wt_type must be UINT, INT, FLOAT, or DOUBLE" AS errMsg;
+		RETURN;
+	END;
+  
+  IF write_size == 0 THEN
+    PRINT "write_size must be positive" AS errMsg;
+    RETURN;
+  END;
+   
+	##### Initialize #####
+	start = {source};
+    start = SELECT s 
+          FROM start:s
+          POST-ACCUM s.@minPath = min_tup(0, 1, s);
+  
+
+	##### Walk one step at a time until there is no more path.
+	WHILE start.size() != 0 DO 
+		start = SELECT t
+			FROM start:s -(e_type:e)-> v_type:t
+			ACCUM 
+          t.@is_candidate = FALSE,
+   
+          CASE wt_type
+            WHEN "UINT" THEN
+			  t.@minPath += min_tup(s.@minPath.top().cost + e.getAttr(wt_attr, "UINT"), 
+                                    s.@minPath.top().length + 1, s)
+            WHEN "INT" THEN
+			  t.@minPath += min_tup(s.@minPath.top().cost + e.getAttr(wt_attr, "INT"), 
+                                    s.@minPath.top().length + 1, s)
+            WHEN "FLOAT" THEN
+              t.@minPath += min_tup(s.@minPath.top().cost + e.getAttr(wt_attr, "FLOAT"), 
+                                    s.@minPath.top().length + 1, s)
+            WHEN "DOUBLE" THEN
+              t.@minPath += min_tup(s.@minPath.top().cost + e.getAttr(wt_attr, "DOUBLE"), 
+                                    s.@minPath.top().length + 1, s)
+          END
+     POST-ACCUM
+       IF abs(t.@prevMinPath - t.@minPath.top().cost) > epsilon THEN
+         t.@is_candidate = TRUE,
+         t.@prevMinPath = t.@minPath.top().cost
+       END
+     HAVING t.@is_candidate;
+
+	END;
+  
+	##### Output #####
+	IF file_path != "" THEN
+	  f.println("Vertex_ID","Distance","Shortest_Path");
+	END;
+  
+  start_set = {source};
+  component = {v_type};
+  component = component MINUS start_set; 
+   
+  component = SELECT s 
+              FROM component:s
+              WHERE s.@prevMinPath != -1
+              POST-ACCUM 
+			          IF result_attr != "" THEN s.setAttr(result_attr, s.@minPath.top().cost) END;
+
+  //store and write paths by batches to reduce memory consumption
```

Comment:
Are we actually using batches?

Reply:
We only use batches for the traceback, which is often very memory consuming. For the other parts, the memory consumed = the local accumulator overhead.

Reply:
Ah, gotcha. Most batch implementations use a FOREACH with a mod of the batch size to randomly sample. I see component MINUS tmp now, though, haha.